File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = 100

    for i in range(n-1):
        arr = wires.copy()
        edge = arr.pop(i)
        visited = [False] * len(arr)
        
        
        # [[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]
        # 여기서 1, 3 dfs
        def dfs(node, arr, count, imsi):
            # print(f"count : {count}")
            for i in range(len(arr)):
                if arr[i][0] == node and not visited[i]:
                    count += 1
                    visited[i] = True
                    imsi.append(count)
                    dfs(arr[i][1], arr, count, imsi)
                
                elif arr[i][1]  == node and not visited[i]:
                    count += 1
                    visited[i] = True
                    imsi.append(count)
                    dfs(arr[i][0], arr, count, imsi)
            # print(f"count2 : {count}")
            return imsi
                    
# 
        
        
        imsi1 = []
        count1 = dfs(edge[0], arr, 0, imsi1)
        imsi2 = []
        count2 = dfs(edge[1], arr, 0, imsi2)
       
        if (answer > abs(len(count1) - len(count2))):
            answer = abs(len(count1) - len(count2))
                        
    return answer


# dfs
# wires 길이만큼 for문
# 각 dfs한 길이의 차 중 가장 적은걸 반환

# 카운트할 변수 2개 설정?
# wires 순서는 상관없다
# for문 안에서 wires

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>
#include <cmath>
using namespace std;


int dfs(int node, vector<int> wire, vector<vector<int>>& graph, vector<bool>& visited){
    int count = 1;
    for(int i: graph[node]){
        if(visited[i] != false){
            continue;
        }
        if((node == wire[0] && i == wire[1] )||(node == wire[1] && i == wire[0])){
            continue;
        }
        visited[i] = true;
        count += dfs(i, wire, graph, visited);
    }
    return count;
}


int solution(int n, vector<vector<int>> wires) {
    int answer = -1;
    // 그래프만들기
    vector<vector<int>> graph(n+1);
    
    for(int i=0; i<wires.size(); i++){
        graph[wires[i][0]].push_back(wires[i][1]);
        graph[wires[i][1]].push_back(wires[i][0]);
    }

    // priority_queue 
    priority_queue<int, vector<int>, greater<int>> pq;
    
    for(int i=0; i<wires.size(); i++){
        // visited
        vector<bool> visited(n+1, false);
        visited[wires[i][0]] = true;
        int count_a = dfs(wires[i][0], wires[i], graph, visited);
        int count_b = abs(n-count_a);
        int total = abs(count_a-count_b);
        pq.push(total);
        
    }
    answer = pq.top();
    return answer;
}


// 인접리스트 만들기 - 그래프 / visited 만들기 n개수대로/ prirority_queue 최소힙
// 1. wires 개수대로 for문
//  - int dfs(구해올 노드, 첫번째 노드두번째 노드 벡터, 그래프, visited)
//  - count_a 값으로 구해올 노드 값 받기


//  - dfs
//  - 현재 노드 개수 count =1


//  - 그래프[현재노드] for문해서
//  
// - 방문한적이 없다면
// - 구해올노드 ==첫번째노드, 그래프[노드] == 두번째노드 or  구해올노드 ==두번째노드, 그래프[노드] == 첫번째노드 라면 continue
//  - count = dfs()
// - 방문처리
// - 마지막에 count반환 

// 2. 받아왔으면 abs(n-count_a)로 두번째 노드값 구하기
// 3. 차를 prirority_queue에 넣음
// 4. 다했으면 미니힙 top확인